[WIP] add server version to `ark version` output

skriss opened the pull request:
Signed-off-by: Steve Kriss <steve@heptio.com>

Fixes #770 

Would appreciate a quick review on the CRD definition before I proceed with the rest of the PR.

cc @heptio/ark-team

carlisia commented:
This lgtm.

ncdc commented on pkg/apis/ark/v1/status_request.go:
```diff
@@ -0,0 +1,71 @@
+	// by the StatusRequestController.
+	ProcessedTimestamp metav1.Time `json:"processedTimestamp"`
+
+	// ServerVersion is the version of the Ark binary that the
```
```suggestion
	// ServerVersion is the Ark server version.
```

skriss commented:
Would appreciate another look at this when folks have a chance now that I've implemented the server-side logic.  I did a couple of things differently than our existing patterns and would like input on whether they're good changes:

- rather than putting all logic directly into the controller, I kept the controller pretty minimal (just enqueueing and dequeuing items) and put the actual domain logic into a new package, `pkg/statusrequest`.
- I added `pkg/builder` with a helper type for building `StatusRequest` objects for test purposes.  it looks pretty similar to some of the builders that are currently in `pkg/util/test` (though I dropped the `With` prefix on every method, e.g. `.Name(...)` instead of `.WithName(...)` to keep things shorter.  Ideally would like to move all builders out of `pkg/util` and into here.

Still need to do the client-side updates to `ark version`.

ncdc commented on pkg/controller/status_request_controller.go:
```diff
@@ -0,0 +1,120 @@
+
+	log.Debug("Getting StatusRequest")
+	req, err := c.lister.StatusRequests(ns).Get(name)
+	if err != nil {
```
I prefer to do these error checks as
```go
if apierrors.IsNotFound(err) {
...
}
if err != nil {
...
}
```

skriss commented:
Added client code.  The output currently looks something like:

```
Client:
        Version: master
        Git commit: v0.10.0-38-gfa40cc19
        Git tree state: dirty
Server:
        Version: v0.10.1
```

ncdc commented:
My bad!

On Thu, Dec 6, 2018 at 5:27 PM Steve Kriss <notifications@github.com> wrote:

> I included one :)
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/heptio/ark/pull/1116#issuecomment-445052830>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAABYiO_4c2wXSq4Fkmu4pJeBt7sHxQQks5u2ZnJgaJpZM4ZDKth>
> .
>

skriss commented:
rebased

skriss commented:
@carlisia @nrb @wwitzel3 @ncdc ready for a full review

ncdc commented on pkg/apis/ark/v1/status_request.go:
```diff
@@ -0,0 +1,70 @@
+
+// StatusRequest is a request to access current status information about
+// the Ark server.
+type StatusRequest struct {
```
Is it worth calling this ServerStatusRequest?

skriss replied:
🤷‍♂️ I think I'm fine either way - strong feelings? Anyone else?

ncdc commented on pkg/apis/ark/v1/status_request.go:
```diff
@@ -0,0 +1,70 @@
+	Phase StatusRequestPhase `json:"phase"`
+
+	// ProcessedTimestamp is when the StatusRequest was processed
+	// by the StatusRequestController.
```
I would omit the controller detail - that's not something an end user would want or need to know about.

ncdc commented on pkg/builder/status_request.go:
```diff
@@ -0,0 +1,74 @@
+limitations under the License.
+*/
+
+package builder
```
Do you anticipate we'll add more builders over time? If so, maybe this organization would make more sense?
```
pkg/builders/
  statusrequest
  someothertype
```

and then in each package, we'd have a plain `Builder` type along with `NewBuilder`, e.g.

```go
package statusrequest

type Builder struct {}

func NewBuilder() *Builder {}
```

We could even align on a common function name to spit out the underlying object - perhaps `Build()`?

ncdc replied:
Or if we're afraid that `pkg/builders/backup` would conflict with `var backup *v1.Backup`, we could suffix all builder packages with `builder` - `pkg/builders/backupbuilder` and then instead of saying `backupbuilder.NewBuilder()` it would just be `backupbuilder.New()`.

skriss replied:
I do expect we'd add more, yes; over time, I'd like this to replace the existing test obj helpers and be a little more succinct and standardized.

What's the motivation for having a subpackage per type?  I'm not sure on that aspect yet.

+1 for renaming to `Build()` to get the obj

ncdc replied:
I was hoping to eliminate the duplication of "builder(s)" in `builders.NewStatusRequestBuilder`.

skriss replied:
maybe it should just go into `pkg/statusrequest`, as a `type Builder struct` and `NewBuilder()` constructor?  And over time as we relocate controller logic into separate packages, each of those can define their own?

ncdc replied:
That seems reasonable. Although we'll have problems with single-word types, such as `backup` and `restore` (trying to use as both a variable name and package name).

ncdc commented on pkg/cmd/version/version.go:
```diff
@@ -18,21 +18,109 @@ package version
 
-func NewCommand() *cobra.Command {
+func NewCommand(f client.Factory) *cobra.Command {
+	timeout := time.Minute
```
I would default this to 3 to 5 seconds

ncdc commented on pkg/cmd/version/version.go:
```diff
@@ -18,21 +18,109 @@ package version
-			fmt.Printf("Version: %s\n", buildinfo.Version)
-			fmt.Printf("Git commit: %s\n", buildinfo.FormattedGitSHA())
+		Run: func(c *cobra.Command, args []string) {
+			fmt.Println("Client:")
```
I know we haven't done this much (if any) to date, but is this worth splitting out into its own function so we can unit test it?

skriss replied:
OK, refactored & added a unit test for what I think is the important part. PTAL.

ncdc commented on pkg/cmd/version/version.go:
```diff
@@ -18,21 +18,109 @@ package version
+	}
+	defer client.StatusRequests(namespace).Delete(created.Name, nil)
+
+	listOptions := metav1.ListOptions{
```
We have this same logic in `downloadrequest.Stream()`. It might be nice in the future to extract this to a helper, with a custom handler function to call in the `watch.Modified` case.

ncdc commented on pkg/statusrequest/process.go:
```diff
@@ -0,0 +1,90 @@
+	arkv1client "github.com/heptio/ark/pkg/generated/clientset/versioned/typed/ark/v1"
+)
+
+const ttl = time.Hour
```
1 minute is probably sufficient, especially if we change the default client-side timeout to 3 or 5 seconds.

ncdc commented on pkg/controller/status_request_controller.go:
```diff
@@ -0,0 +1,120 @@
+	}
+
+	for _, req := range items {
+		if req.Status.Phase == arkv1api.StatusRequestPhaseProcessed {
```
Is it worth readding everything, not just processed ones?

ncdc commented on pkg/cmd/version/version.go:
```diff
@@ -18,21 +18,133 @@ package version
+
+			switch e.Type {
+			case watch.Deleted:
+				return nil, errors.New("download request was unexpectedly deleted")
```
download -> server status

ncdc commented on pkg/cmd/version/version.go:
```diff
@@ -18,21 +18,133 @@ package version
+			case watch.Deleted:
+				return nil, errors.New("download request was unexpectedly deleted")
+			case watch.Modified:
+				if updated.Status.Phase == arkv1api.ServerStatusRequestPhaseProcessed && updated.Status.ServerVersion != "" {
```
Don't we want to break out of the loop if the phase is processed, regardless of the value of `updated.Status.ServerVersion`?

nrb commented:
Could you include a changelog entry?

skriss commented:
@ncdc could you take another look here? your comments should all be addressed. will squash before merge.

ncdc commented on pkg/apis/ark/v1/register.go:
```diff
@@ -69,6 +69,7 @@ func CustomResources() map[string]typeInfo {
 		"ResticRepository":       newTypeInfo("resticrepositories", &ResticRepository{}, &ResticRepositoryList{}),
 		"BackupStorageLocation":  newTypeInfo("backupstoragelocations", &BackupStorageLocation{}, &BackupStorageLocationList{}),
 		"VolumeSnapshotLocation": newTypeInfo("volumesnapshotlocations", &VolumeSnapshotLocation{}, &VolumeSnapshotLocationList{}),
+		"ServerStatusRequest":    newTypeInfo("statusrequests", &ServerStatusRequest{}, &ServerStatusRequestList{}),
```
Does it need to be this?

```suggestion
		"ServerStatusRequest":    newTypeInfo("serverstatusrequests", &ServerStatusRequest{}, &ServerStatusRequestList{}),
```

ncdc commented on pkg/controller/server_status_request_controller.go:
```diff
@@ -0,0 +1,118 @@
+				}).Debug("Enqueueing server status request")
+
+				c.queue.Add(key)
+			},
```
I think it's probably worth having an `UpdateFunc` here too - if it switches from New to Processed and for whatever reason the client-side delete request misfires, the `UpdateFunc` would trigger a quick controller-initiated delete, instead of waiting up to an hour for all items to be reenqueued. WDYT?

ncdc replied:
Oh wait, but the server side logic waits for the expiration, so it won't be expired yet necessarily

skriss replied:
I could just drop the resync period down to something smaller (5 min?) so if anything does get stranded, it's cleaned up more quickly

ncdc replied:
Yes, let's do that

ncdc commented:
@skriss did another pass.

skriss commented:
Changes made. Anything else? Will do some retesting this aft as well.

ncdc commented:
LGTM
On Jan 23, 2019, 1:34 PM -0500, Steve Kriss <notifications@github.com>, wrote:

Changes made. Anything else? Will do some retesting this aft as well.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fheptio%2Fark%2Fpull%2F1116%23issuecomment-456915937&data=02%7C01%7Cgoldsteina%40vmware.com%7C42644268e32249b28e0208d681615b2f%7Cb39138ca3cee4b4aa4d6cd83d9dd62f0%7C0%7C0%7C636838652478476983&sdata=PeWwG7wPCljATO1EENwa%2F7yjWxn0MdSnoeTfQYxKvEQ%3D&reserved=0>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAAABYs6wAZMYP5fqgUnPRpEp_ctADQFVks5vGKsdgaJpZM4ZDKth&data=02%7C01%7Cgoldsteina%40vmware.com%7C42644268e32249b28e0208d681615b2f%7Cb39138ca3cee4b4aa4d6cd83d9dd62f0%7C0%7C0%7C636838652478487002&sdata=%2FS%2BN2kcKeN26IadgRjbzXpBRufII%2Beopeh9LiDNFgok%3D&reserved=0>.

skriss commented:
tests LGTM.

skriss commented:
squished.
